subsys/bluetooth/mesh/adv.c

```diff
@@ -41,88 +41,157 @@ const uint8_t bt_mesh_adv_type[BT_MESH_ADV_TYPES] = {
-						    struct bt_mesh_adv *adv_pool,
-						    enum bt_mesh_adv_type type,
-						    enum bt_mesh_adv_tag tag,
-						    uint8_t xmit, k_timeout_t timeout)
+static struct bt_mesh_buf *adv_create_from_pool(struct k_mem_slab *buf_pool,
+						enum bt_mesh_adv_type type,
+						enum bt_mesh_adv_tag tag,
+						uint8_t xmit, uint8_t prio,
+						k_timeout_t timeout)
 {
 	struct bt_mesh_adv *adv;
-	struct net_buf *buf;
+	struct bt_mesh_buf *buf;
+	int err;
 
 	if (atomic_test_bit(bt_mesh.flags, BT_MESH_SUSPENDED)) {
 		LOG_WRN("Refusing to allocate buffer while suspended");
 		return NULL;
 	}
 
-	buf = net_buf_alloc(buf_pool, timeout);
-	if (!buf) {
+	err = k_mem_slab_alloc(buf_pool, (void **)&buf, timeout);
+	if (err) {
 		return NULL;
 	}
 
-	adv = &adv_pool[net_buf_id(buf)];
-	BT_MESH_ADV(buf) = adv;
+	buf->ref = 1;
+
+	net_buf_simple_init_with_data(&buf->b, buf->_bufs, BT_MESH_ADV_DATA_SIZE);
+	net_buf_simple_reset(&buf->b);
+
+	adv = &buf->adv;
 
 	(void)memset(adv, 0, sizeof(*adv));
 
 	adv->type         = type;
 	adv->tag          = tag;
 	adv->xmit         = xmit;
+	adv->prio         = prio;
 
 	return buf;
 }
 
-struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type,
-				   enum bt_mesh_adv_tag tag,
-				   uint8_t xmit, k_timeout_t timeout)
+struct bt_mesh_buf *bt_mesh_buf_ref(struct bt_mesh_buf *buf)
 {
+	buf->ref++;
+
+	return buf;
+}
+
+void bt_mesh_buf_unref(struct bt_mesh_buf *buf)
+{
+	if (--buf->ref > 0) {
+		return;
+	}
+
+	struct k_mem_slab *slab = &adv_buf_pool;
+	struct bt_mesh_adv adv = *(&buf->adv);
+
 #if defined(CONFIG_BT_MESH_RELAY)
-	if (tag & BT_MESH_RELAY_ADV) {
-		return bt_mesh_adv_create_from_pool(&relay_buf_pool,
-						    adv_relay_pool, type,
-						    tag, xmit, timeout);
+	if (adv.tag == BT_MESH_RELAY_ADV) {
+		slab = &relay_buf_pool;
 	}
-#endif
+#endif /* CONFIG_BT_MESH_RELAY */
 
-	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_local_pool, type,
-					    tag, xmit, timeout);
+	k_mem_slab_free(slab, (void **)&buf);
+	bt_mesh_adv_send_end(0, &adv);
+}
+
+struct bt_mesh_buf *bt_mesh_adv_relay_create(uint8_t prio, uint8_t xmit)
+{
+#if defined(CONFIG_BT_MESH_RELAY)
+	void *lowest = NULL, *prev = NULL, *curr;
+	uint8_t prio_cur = prio;
+	struct bt_mesh_buf *buf;
+
+	buf = adv_create_from_pool(&relay_buf_pool,
+				   BT_MESH_ADV_DATA, BT_MESH_RELAY_ADV,
+				   xmit, prio, K_NO_WAIT);
+	if (buf) {
+		return buf;
+	}
+
+	if (!IS_ENABLED(CONFIG_BT_MESH_RELAY_PRIORITY) || !prio_cur) {
+		return NULL;
+	}
+
+	K_QUEUE_FOR_EACH(&bt_mesh_relay_queue, curr) {
+		buf = CONTAINER_OF(curr, struct bt_mesh_buf, node);
+
+		if (buf->adv.prio < prio_cur) {
+			prio_cur = buf->adv.prio;
+			lowest = curr;
+		}
+
+		prev = curr;
+	}
+
+	if (!lowest) {
+		return NULL;
+	}
+
+	k_queue_remove(&bt_mesh_relay_queue, lowest);
+
+	buf = CONTAINER_OF(lowest, struct bt_mesh_buf, node);
+
+	bt_mesh_adv_send_start(0, -ECANCELED, &buf->adv);
+	bt_mesh_buf_unref(buf);
```

Comment:
For first question:

Reply:
Right, I've missed that, thanks! The first point is clear now.

Reply:
For second question:
When sending a message in adv_ext.c, will dequeue from queue, so this adv buf will not be trace here.
And
For proxy_msg.c, good point, i'm missing this condition

Reply:
I thought about it carefully, the second problem you mentioned about proxy relay, the final reason may be that we confuse the concept of relay, in the profile of Mesh1.0, relay in many scenarios is defined, I think this is not Should not be addressed in this PR.

Reply:
### Mesh Profile 1.0.1 3.4.6.3 Receiving a Network PDU

If the message delivered from the advertising bearer is processed by the lower transport layer, the Relay
feature is supported and enabled, the TTL field has a value of 2 or greater, and the destination is not a
unicast address of an element on this node, then the TTL field value shall be decremented by 1, the
Network PDU shall be **tagged as relay**, and the Network PDU shall be retransmitted to all network
interfaces connected to the advertising bearer. It is recommended that a small random delay is introduced
between receiving a Network PDU and relaying a Network PDU to avoid collisions between multiple
relays that have received the Network PDU at the same time.

If the message delivered from the GATT bearer is processed by the lower transport layer, and the Proxy
feature is supported and enabled, and the TTL field has a value of 2 or greater, and the destination is not
a unicast address of an element on this node, then the TTL field value shall be decremented by 1, and
the Network PDU shall be retransmitted to all network interfaces.

If the message delivered from the advertising bearer is processed by the lower transport layer, and the
Proxy feature is supported and enabled, and the TTL field has a value of 2 or greater, and the destination
is not a unicast address of an element on this node, then the TTL field shall be decremented by 1 and the
Network PDU shall be retransmitted to all network interfaces connected to the GATT bearer.

Reply:
Also done now.

Reply:
@PavelVPV  All comments have been completed. Please take a moment to have a look. I don't want to put it on hold for too long, otherwise have to do rebase later.

Reply:
Sorry, a bit busy at the moment. Anyway, I'd like @alxelax and @omkar3141 to look at this PR before we merge it, but they will be back in January.

Reply:
I didn't get your point. If the node has a proxy client connected to it and GATT Proxy feature is enabled, the non-filtered out messages will be relayed. When such node has prioritization enabled, this code will try to release the space in the pool, but will fail because of the additional reference in `proxy_msg.c`. So what will happen then is that such node will relay messages to the proxy client, but not over adv bearer.

Reply:
please see my current implementation, the proxy will use separate buffer.

Reply:
BTW： the proxy and friend relay message will use adv main pool instead of relay pool.